In [1]:
import pandas as pd
import numpy as np

# Procesado de los datos

## Lectura de datos

In [2]:
df = pd.read_csv("flights.csv")
pd.set_option('display.max_columns', None)
df.head()

C:\Users\gfine\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3326: DtypeWarning: Columns (7,8) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0  2015      1    1            4      AS             98      N407AS   
1  2015      1    1            4      AA           2336      N3KUAA   
2  2015      1    1            4      US            840      N171US   
3  2015      1    1            4      AA            258      N3HYAA   
4  2015      1    1            4      AS            135      N527AS   

  ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  DEPARTURE_TIME  \
0            ANC                 SEA                    5          2354.0   
1            LAX                 PBI                   10             2.0   
2            SFO                 CLT                   20            18.0   
3            LAX                 MIA                   20            15.0   
4            SEA                 ANC                   25            24.0   

   DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  \
0            -11.0      21.0        15.0           205.0         194.0   
1             -8.0      12.0        14.0           280.0         279.0   
2             -2.0      16.0        34.0           286.0         293.0   
3             -5.0      15.0        30.0           285.0         281.0   
4             -1.0      11.0        35.0           235.0         215.0   

   AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  SCHEDULED_ARRIVAL  ARRIVAL_TIME  \
0     169.0      1448      404.0      4.0                430         408.0   
1     263.0      2330      737.0      4.0                750         741.0   
2     266.0      2296      800.0     11.0                806         811.0   
3     258.0      2342      748.0      8.0                805         756.0   
4     199.0      1448      254.0      5.0                320         259.0   

   ARRIVAL_DELAY  DIVERTED  CANCELLED CANCELLATION_REASON  AIR_SYSTEM_DELAY  \
0          -22.0         0          0                 NaN               NaN   
1           -9.0         0          0                 NaN               NaN   
2            5.0         0          0                 NaN               NaN   
3           -9.0         0          0                 NaN               NaN   
4          -21.0         0          0                 NaN               NaN   

   SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0             NaN            NaN                  NaN            NaN  
1             NaN            NaN                  NaN            NaN  
2             NaN            NaN                  NaN            NaN  
3             NaN            NaN                  NaN            NaN  
4             NaN            NaN                  NaN            NaN

In [3]:
# Eliminamos las filas duplicadas. Un mismo vuelo no puede retrasarse en el mismo momento del tiempo más de una vez
df = df.drop_duplicates()
len(df)

5819079

In [6]:
airports_df = pd.read_csv("airports.csv")
airports_df.head()

IATA_CODE                              AIRPORT         CITY STATE COUNTRY  \
0       ABE  Lehigh Valley International Airport    Allentown    PA     USA   
1       ABI             Abilene Regional Airport      Abilene    TX     USA   
2       ABQ    Albuquerque International Sunport  Albuquerque    NM     USA   
3       ABR            Aberdeen Regional Airport     Aberdeen    SD     USA   
4       ABY   Southwest Georgia Regional Airport       Albany    GA     USA   

   LATITUDE  LONGITUDE  
0  40.65236  -75.44040  
1  32.41132  -99.68190  
2  35.04022 -106.60919  
3  45.44906  -98.42183  
4  31.53552  -84.19447

In [10]:
airports = airports_df['IATA_CODE']
pocho = df[df.ORIGIN_AIRPORT.isin(airports)==False]
pocho.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
4385712  2015     10    1            4      AA           1230      N3DBAA   
4385713  2015     10    1            4      DL           1805      N696DL   
4385714  2015     10    1            4      NK            612      N602NK   
4385715  2015     10    1            4      AA            260      N3GNAA   
4385716  2015     10    1            4      AA           1982      N914UY   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  \
4385712          14747               11298                    5   
4385713          14771               13487                    5   
4385714          12889               13487                    5   
4385715          12892               13303                   10   
4385716          14771               11057                   10   

         DEPARTURE_TIME  DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  \
4385712            15.0             10.0      15.0        30.0   
4385713            16.0             11.0      14.0        30.0   
4385714          2400.0             -5.0      15.0        15.0   
4385715             7.0             -3.0      28.0        35.0   
4385716             8.0             -2.0      12.0        20.0   

         SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  \
4385712           237.0         208.0     189.0      1660      539.0      4.0   
4385713           213.0         192.0     171.0      1589      521.0      7.0   
4385714           177.0         168.0     149.0      1299      444.0      4.0   
4385715           296.0         306.0     271.0      2342      806.0      7.0   
4385716           291.0         282.0     257.0      2296      737.0     13.0   

         SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
4385712                602         543.0          -19.0         0          0   
4385713                538         528.0          -10.0         0          0   
4385714                502         448.0          -14.0         0          0   
4385715                806         813.0            7.0         0          0   
4385716                801         750.0          -11.0         0          0   

        CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
4385712                 NaN               NaN             NaN            NaN   
4385713                 NaN               NaN             NaN            NaN   
4385714                 NaN               NaN             NaN            NaN   
4385715                 NaN               NaN             NaN            NaN   
4385716                 NaN               NaN             NaN            NaN   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
4385712                  NaN            NaN  
4385713                  NaN            NaN  
4385714                  NaN            NaN  
4385715                  NaN            NaN  
4385716                  NaN            NaN

In [11]:
pocho['MONTH'].unique()

array([10], dtype=int64)

In [19]:
df['MONTH'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12], dtype=int64)

In [30]:
dg = pd.DataFrame(df.groupby(df['MONTH'])["FLIGHT_NUMBER"].count())
dg = dg.rename(columns={"FLIGHT_NUMBER":"TOTAL_FLIGHTS"})
dg

TOTAL_FLIGHTS
MONTH               
1             469968
2             429191
3             504312
4             485151
5             496993
6             503897
7             520718
8             510536
9             464946
10            486165
11            467972
12            479230

In [33]:
is_oct = df.loc[:, 'MONTH'] == 10
df_oct = df.loc[is_oct]
df_oct.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
4385712  2015     10    1            4      AA           1230      N3DBAA   
4385713  2015     10    1            4      DL           1805      N696DL   
4385714  2015     10    1            4      NK            612      N602NK   
4385715  2015     10    1            4      AA            260      N3GNAA   
4385716  2015     10    1            4      AA           1982      N914UY   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  \
4385712          14747               11298                    5   
4385713          14771               13487                    5   
4385714          12889               13487                    5   
4385715          12892               13303                   10   
4385716          14771               11057                   10   

         DEPARTURE_TIME  DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  \
4385712            15.0             10.0      15.0        30.0   
4385713            16.0             11.0      14.0        30.0   
4385714          2400.0             -5.0      15.0        15.0   
4385715             7.0             -3.0      28.0        35.0   
4385716             8.0             -2.0      12.0        20.0   

         SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  \
4385712           237.0         208.0     189.0      1660      539.0      4.0   
4385713           213.0         192.0     171.0      1589      521.0      7.0   
4385714           177.0         168.0     149.0      1299      444.0      4.0   
4385715           296.0         306.0     271.0      2342      806.0      7.0   
4385716           291.0         282.0     257.0      2296      737.0     13.0   

         SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
4385712                602         543.0          -19.0         0          0   
4385713                538         528.0          -10.0         0          0   
4385714                502         448.0          -14.0         0          0   
4385715                806         813.0            7.0         0          0   
4385716                801         750.0          -11.0         0          0   

        CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
4385712                 NaN               NaN             NaN            NaN   
4385713                 NaN               NaN             NaN            NaN   
4385714                 NaN               NaN             NaN            NaN   
4385715                 NaN               NaN             NaN            NaN   
4385716                 NaN               NaN             NaN            NaN   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
4385712                  NaN            NaN  
4385713                  NaN            NaN  
4385714                  NaN            NaN  
4385715                  NaN            NaN  
4385716                  NaN            NaN

In [35]:
df['ORIGIN_AIRPORT'].unique()

array(['ANC', 'LAX', 'SFO', 'SEA', 'LAS', 'DEN', 'SLC', 'PDX', 'FAI',
       'MSP', 'PHX', 'SJU', 'PBG', 'IAG', 'PSE', 'BQN', 'ORD', 'GEG',
       'HNL', 'ONT', 'MCO', 'BOS', 'HIB', 'ABR', 'MAF', 'DFW', 'MKE',
       'IAH', 'BNA', 'BRO', 'VPS', 'BOI', 'BJI', 'SGF', 'PHL', 'SBN',
       'RDD', 'EUG', 'IAD', 'BUF', 'PWM', 'JFK', 'CRP', 'PIA', 'FAT',
       'SMF', 'AUS', 'MCI', 'ATL', 'JAX', 'MFR', 'IDA', 'MSN', 'DCA',
       'SAT', 'CHS', 'SBA', 'SMX', 'IND', 'CLE', 'GSP', 'BDL', 'ABI',
       'RIC', 'BFL', 'OMA', 'RDM', 'FLL', 'CID', 'TPA', 'SYR', 'ROC',
       'TYR', 'LAN', 'XNA', 'GSO', 'EWR', 'PBI', 'RSW', 'OAK', 'PVD',
       'RNO', 'PIT', 'ABQ', 'MIA', 'BWI', 'LGA', 'TUL', 'LIT', 'MSY',
       'OKC', 'ATW', 'PNS', 'MEM', 'TYS', 'MHT', 'SAV', 'CLT', 'GRB',
       'ABE', 'JAN', 'OAJ', 'FAR', 'ERI', 'LEX', 'CWA', 'MSO', 'TTN',
       'AMA', 'CLL', 'HOU', 'JLN', 'MLI', 'RDU', 'CVG', 'MHK', 'MOB',
       'TLH', 'BHM', 'CAE', 'TXK', 'ACY', 'DTW', 'RAP', 'TUS', 'EAU',
       'DLH', 'FSD',

## Selección de los datos

Trabajaremos únicamente con los datos correspondientes a los 10 mayores aeropuertos de Estados Unidos de acuerdo con:

https://www.datosmundial.com/america/usa/aeropuertos.php

In [22]:
# Lista con el top 10 de aeropuertos
top_10_airports = ['JFK','LAX','ORD','MIA','SFO','IAD','ATL','EWR','MCO','BOS','LAS','SEA','MSY','DFW','IAH']

# Filtramos los datos
df = df[df.ORIGIN_AIRPORT.isin(top_10_airports)]
len(df)

2162516

In [23]:
dg = pd.DataFrame(df.groupby(df['MONTH'])["FLIGHT_NUMBER"].count())
dg = dg.rename(columns={"FLIGHT_NUMBER":"TOTAL_FLIGHTS"})
dg

TOTAL_FLIGHTS
MONTH               
1             189598
2             173541
3             202656
4             195948
5             202538
6             205338
7             211142
8             208394
9             190077
11            189406
12            193878

In [9]:
df['MONTH'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 11, 12], dtype=int64)

In [10]:
df.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
1  2015      1    1            4      AA           2336      N3KUAA   
2  2015      1    1            4      US            840      N171US   
3  2015      1    1            4      AA            258      N3HYAA   
5  2015      1    1            4      DL            806      N3730B   
7  2015      1    1            4      US           2013      N584UW   

  ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  DEPARTURE_TIME  \
1            LAX                 PBI                   10             2.0   
2            SFO                 CLT                   20            18.0   
3            LAX                 MIA                   20            15.0   
5            SFO                 MSP                   25            20.0   
7            LAX                 CLT                   30            44.0   

   DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  \
1             -8.0      12.0        14.0           280.0         279.0   
2             -2.0      16.0        34.0           286.0         293.0   
3             -5.0      15.0        30.0           285.0         281.0   
5             -5.0      18.0        38.0           217.0         230.0   
7             14.0      13.0        57.0           273.0         249.0   

   AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  SCHEDULED_ARRIVAL  ARRIVAL_TIME  \
1     263.0      2330      737.0      4.0                750         741.0   
2     266.0      2296      800.0     11.0                806         811.0   
3     258.0      2342      748.0      8.0                805         756.0   
5     206.0      1589      604.0      6.0                602         610.0   
7     228.0      2125      745.0      8.0                803         753.0   

   ARRIVAL_DELAY  DIVERTED  CANCELLED CANCELLATION_REASON  AIR_SYSTEM_DELAY  \
1           -9.0         0          0                 NaN               NaN   
2            5.0         0          0                 NaN               NaN   
3           -9.0         0          0                 NaN               NaN   
5            8.0         0          0                 NaN               NaN   
7          -10.0         0          0                 NaN               NaN   

   SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
1             NaN            NaN                  NaN            NaN  
2             NaN            NaN                  NaN            NaN  
3             NaN            NaN                  NaN            NaN  
5             NaN            NaN                  NaN            NaN  
7             NaN            NaN                  NaN            NaN

## Análisis de variables

In [5]:
df.columns

Index(['YEAR', 'MONTH', 'DAY', 'DAY_OF_WEEK', 'AIRLINE', 'FLIGHT_NUMBER',
       'TAIL_NUMBER', 'ORIGIN_AIRPORT', 'DESTINATION_AIRPORT',
       'SCHEDULED_DEPARTURE', 'DEPARTURE_TIME', 'DEPARTURE_DELAY', 'TAXI_OUT',
       'WHEELS_OFF', 'SCHEDULED_TIME', 'ELAPSED_TIME', 'AIR_TIME', 'DISTANCE',
       'WHEELS_ON', 'TAXI_IN', 'SCHEDULED_ARRIVAL', 'ARRIVAL_TIME',
       'ARRIVAL_DELAY', 'DIVERTED', 'CANCELLED', 'CANCELLATION_REASON',
       'AIR_SYSTEM_DELAY', 'SECURITY_DELAY', 'AIRLINE_DELAY',
       'LATE_AIRCRAFT_DELAY', 'WEATHER_DELAY'],
      dtype='object')

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1493459 entries, 1 to 5819078
Data columns (total 31 columns):
 #   Column               Non-Null Count    Dtype  
---  ------               --------------    -----  
 0   YEAR                 1493459 non-null  int64  
 1   MONTH                1493459 non-null  int64  
 2   DAY                  1493459 non-null  int64  
 3   DAY_OF_WEEK          1493459 non-null  int64  
 4   AIRLINE              1493459 non-null  object 
 5   FLIGHT_NUMBER        1493459 non-null  int64  
 6   TAIL_NUMBER          1488330 non-null  object 
 7   ORIGIN_AIRPORT       1493459 non-null  object 
 8   DESTINATION_AIRPORT  1493459 non-null  object 
 9   SCHEDULED_DEPARTURE  1493459 non-null  int64  
 10  DEPARTURE_TIME       1468925 non-null  float64
 11  DEPARTURE_DELAY      1468925 non-null  float64
 12  TAXI_OUT             1468135 non-null  float64
 13  WHEELS_OFF           1468135 non-null  float64
 14  SCHEDULED_TIME       1493457 non-null  float64
 15

Columnas que se pueden eliminar:
- TAIL NUMBER: representa un ID de avión único por lo que no corresponde con una variable representativa para el análisis por lo que la eliminamos

### Debatir:
Nos importan..?
- SCHEDULED_DEPARTURE: yo diría que no, si llega en hora = OK!
- DEPARTURE_TIME: es interesante analizar los retrasos por tramo horario?
- DEPARTURE_DELAY: yo diría que no, si sale tarde pero llega en hora = OK!
- ARRIVAL_TIME: yo diría que no, solo me interesa si el vuelo llega tarde, no?
- SCHEDULED_TIME: no se a que se refiere  
- SCHEDULED_ARRIVAL: solo nos interesa saber si se retrasa o no, no la previsión

In [7]:
# Eliminamos tail number porque es un identificador de avión (no vuelo) único
df = df.drop(["TAIL_NUMBER"],axis=1)

Hay un error de formato con la variable flight number, que es un indicador del vuelo y por lo tanto una variable categórica. Lo mismo ocurre con Cancelled y Diverted

In [8]:
# Añadimos el FlightNum, Cancelled and Diverted como variables categóricas
df['FLIGHT_NUMBER']=df['FLIGHT_NUMBER'].astype(object) 
df['CANCELLED']=df['CANCELLED'].astype(object) 
df['DIVERTED']=df['DIVERTED'].astype(object) 

In [9]:
# Unimos las columnas Year, Month y Day of Month como una única variable fecha
# Formato por defecto mes/dia/año para que to_datetime funcione correctamente
df["DATE"]  = df['MONTH'].astype(str) +'/'+ df['DAY'].astype(str) +'/' + df['YEAR'].astype(str)
df["DATE"] = pd.to_datetime(df["DATE"])

Si nos interesa en algún momento podríamos añadir las horas también 

In [10]:
#Eliminamos las columnas year, month y day
df = df.drop(["YEAR","MONTH", "DAY"],axis=1)

# También podemos eliminar la columna DAY_OF_WEEK ya que podemos obtenerla
df = df.drop("DAY_OF_WEEK",axis=1)
print(f"Ejemplo: obtener dia de la semana de la fecha {df['DATE'].iloc[0]} --> {df['DATE'].iloc[0].day_name()} (dia {df['DATE'].iloc[0].dayofweek})")

# NOTA: hay que tener en cuenta que dayofweek empieza a contar en 0 = lunes

Ejemplo: obtener dia de la semana de la fecha 2015-01-01 00:00:00 --> Thursday (dia 3)


In [11]:
# Cambiamos el orden de las columnas, para que DATE sea la primera
cols = df.columns.tolist()
cols = cols[-1:] + cols[:-1]
df = df[cols] 
df.head()

DATE AIRLINE FLIGHT_NUMBER ORIGIN_AIRPORT DESTINATION_AIRPORT  \
1 2015-01-01      AA          2336            LAX                 PBI   
2 2015-01-01      US           840            SFO                 CLT   
3 2015-01-01      AA           258            LAX                 MIA   
5 2015-01-01      DL           806            SFO                 MSP   
7 2015-01-01      US          2013            LAX                 CLT   

   SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  \
1                   10             2.0             -8.0      12.0        14.0   
2                   20            18.0             -2.0      16.0        34.0   
3                   20            15.0             -5.0      15.0        30.0   
5                   25            20.0             -5.0      18.0        38.0   
7                   30            44.0             14.0      13.0        57.0   

   SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  \
1           280.0         279.0     263.0      2330      737.0      4.0   
2           286.0         293.0     266.0      2296      800.0     11.0   
3           285.0         281.0     258.0      2342      748.0      8.0   
5           217.0         230.0     206.0      1589      604.0      6.0   
7           273.0         249.0     228.0      2125      745.0      8.0   

   SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY DIVERTED CANCELLED  \
1                750         741.0           -9.0        0         0   
2                806         811.0            5.0        0         0   
3                805         756.0           -9.0        0         0   
5                602         610.0            8.0        0         0   
7                803         753.0          -10.0        0         0   

  CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
1                 NaN               NaN             NaN            NaN   
2                 NaN               NaN             NaN            NaN   
3                 NaN               NaN             NaN            NaN   
5                 NaN               NaN             NaN            NaN   
7                 NaN               NaN             NaN            NaN   

   LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
1                  NaN            NaN  
2                  NaN            NaN  
3                  NaN            NaN  
5                  NaN            NaN  
7                  NaN            NaN

Hablar formato tiempos: 
    SCHEDULED_DEPARTURE, DEPARTURE_TIME, SCHEDULED_TIME
    SCHEDULED_ARRIVAL, ARRIVAL_TIME (se podrían eliminar dado que nos sirven para sacar arrival delay)

## Debatir:

Las variables SCHEDULED_DEPARTURE, DEPARTURE_TIME, DEPARTURE_DELAY, ARRIVAL_TIME y SCHEDULED_ARRIVAL nos ayudan a saber si un vuelo se ha retrasado o no, pero una vez aue sabemos esto no tienen  valor en sí mismas por lo que las podemos eliminar

#### Tipos de vuelo:
1. On time/ arrived earlier --> arrival_delay <=0
2. Delayed   --> arrival_delay > 0
3. Diverted  --> diverted == 1
4. Cancelled --> cancelled == 1

##### Razones por las que se retrasa un vuelo:
- AIR_SYSTEM_DELAY     
- SECURITY_DELAY       
- AIRLINE_DELAY        
- LATE_AIRCRAFT_DELAY 
- WEATHER_DELAY       

In [12]:
# Vamos a examinar ahora los valores nulos
df.isna().sum()

DATE                         0
AIRLINE                      0
FLIGHT_NUMBER                0
ORIGIN_AIRPORT               0
DESTINATION_AIRPORT          0
SCHEDULED_DEPARTURE          0
DEPARTURE_TIME           24534
DEPARTURE_DELAY          24534
TAXI_OUT                 25324
WHEELS_OFF               25324
SCHEDULED_TIME               2
ELAPSED_TIME             29600
AIR_TIME                 29600
DISTANCE                     0
WHEELS_ON                26273
TAXI_IN                  26273
SCHEDULED_ARRIVAL            0
ARRIVAL_TIME             26273
ARRIVAL_DELAY            29600
DIVERTED                     0
CANCELLED                    0
CANCELLATION_REASON    1467900
AIR_SYSTEM_DELAY       1192588
SECURITY_DELAY         1192588
AIRLINE_DELAY          1192588
LATE_AIRCRAFT_DELAY    1192588
WEATHER_DELAY          1192588
dtype: int64

Las variables CANCELLATION_REASON, AIR_SYSTEM_DELAY, SECURITY_DELAY, AIRLINE_DELAY, LATE_AIRCRAFT_DELAY,WEATHER_DELAY presentan una gran cantidad de valores nulos. No obstante, todas estas variables se relacionan con vuelos cancelados o retrasados por lo que tiene sentido que sean valores nulos para aquellos vuelos que no se hayan cancelado ni retrasado. Vamos a analizar estas variables

In [13]:
cancelled = df[df['CANCELLED'] == 1]
cancelled.isna().sum()

DATE                       0
AIRLINE                    0
FLIGHT_NUMBER              0
ORIGIN_AIRPORT             0
DESTINATION_AIRPORT        0
SCHEDULED_DEPARTURE        0
DEPARTURE_TIME         24534
DEPARTURE_DELAY        24534
TAXI_OUT               25324
WHEELS_OFF             25324
SCHEDULED_TIME             2
ELAPSED_TIME           25559
AIR_TIME               25559
DISTANCE                   0
WHEELS_ON              25559
TAXI_IN                25559
SCHEDULED_ARRIVAL          0
ARRIVAL_TIME           25559
ARRIVAL_DELAY          25559
DIVERTED                   0
CANCELLED                  0
CANCELLATION_REASON        0
AIR_SYSTEM_DELAY       25559
SECURITY_DELAY         25559
AIRLINE_DELAY          25559
LATE_AIRCRAFT_DELAY    25559
WEATHER_DELAY          25559
dtype: int64

Vemos como en este caso no existen valores nulos para la columna CANCELLATION_REASON. No obstante, nuestro análisis consiste en prededcir el retraso de vuelos por lo que no necesitamos la info de vuelos cancelados/ redirigidos así que eliminamos dichos registros

In [14]:
# Eliminamos los cancelados
df = df[df['CANCELLED'] == 0]

#Eliminamos los redirigidos
df = df[df['DIVERTED'] == 0]

df = df.drop(["DIVERTED","CANCELLED", "CANCELLATION_REASON"],axis=1)

Nos quedamos solo con los vuelos que llegan antes de lo previsto, en hora o con retraso

In [15]:
len(df)

1463859

In [16]:
df.isna().sum()

DATE                         0
AIRLINE                      0
FLIGHT_NUMBER                0
ORIGIN_AIRPORT               0
DESTINATION_AIRPORT          0
SCHEDULED_DEPARTURE          0
DEPARTURE_TIME               0
DEPARTURE_DELAY              0
TAXI_OUT                     0
WHEELS_OFF                   0
SCHEDULED_TIME               0
ELAPSED_TIME                 0
AIR_TIME                     0
DISTANCE                     0
WHEELS_ON                    0
TAXI_IN                      0
SCHEDULED_ARRIVAL            0
ARRIVAL_TIME                 0
ARRIVAL_DELAY                0
AIR_SYSTEM_DELAY       1162988
SECURITY_DELAY         1162988
AIRLINE_DELAY          1162988
LATE_AIRCRAFT_DELAY    1162988
WEATHER_DELAY          1162988
dtype: int64

Realizamos ahora el análisis de los valores nulos relacionados con los vuelos retrasados.

**NOTA**: consideramos que un vuelo se retrasa si llega pasada la hora prevista, independientemente de si ha tenido retraso en la hora de salida o no

In [17]:
delayed = df[df['ARRIVAL_DELAY'] > 0]
delayed.isna().sum()

DATE                        0
AIRLINE                     0
FLIGHT_NUMBER               0
ORIGIN_AIRPORT              0
DESTINATION_AIRPORT         0
SCHEDULED_DEPARTURE         0
DEPARTURE_TIME              0
DEPARTURE_DELAY             0
TAXI_OUT                    0
WHEELS_OFF                  0
SCHEDULED_TIME              0
ELAPSED_TIME                0
AIR_TIME                    0
DISTANCE                    0
WHEELS_ON                   0
TAXI_IN                     0
SCHEDULED_ARRIVAL           0
ARRIVAL_TIME                0
ARRIVAL_DELAY               0
AIR_SYSTEM_DELAY       252515
SECURITY_DELAY         252515
AIRLINE_DELAY          252515
LATE_AIRCRAFT_DELAY    252515
WEATHER_DELAY          252515
dtype: int64

Si considerabamos todos los vuelos teníamos 1162988 registros con valor nulo en las variables DELAY, cuando filtramos por vuelos retrasados tenemos 252515.
Vamos a analizar primero los valores nulos de aquellos vuelos que se han retrasado

In [18]:
delayed = delayed[['AIR_SYSTEM_DELAY', 'SECURITY_DELAY','AIRLINE_DELAY','LATE_AIRCRAFT_DELAY','WEATHER_DELAY']]
delayed.head()

AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
2                NaN             NaN            NaN                  NaN   
5                NaN             NaN            NaN                  NaN   
56               NaN             NaN            NaN                  NaN   
63               NaN             NaN            NaN                  NaN   
71               NaN             NaN            NaN                  NaN   

    WEATHER_DELAY  
2             NaN  
5             NaN  
56            NaN  
63            NaN  
71            NaN

In [19]:
# Seleccionamos todas las filas que tengan NaN en todas las columnas
nulls = delayed.loc[(delayed['AIR_SYSTEM_DELAY'].isnull() == True) & (delayed['SECURITY_DELAY'].isnull() == True) & (delayed['AIRLINE_DELAY'].isnull() == True) & (delayed['LATE_AIRCRAFT_DELAY'].isnull() == True) & (delayed['WEATHER_DELAY'].isnull() == True)]
len(nulls)

252515

Vemos que TODOS los NaN se concentran en las mismas filas. Entendemos que en este caso el vuelo se ha retrasado por causa desconocida. Para indicar esto creamos una nueva columna 'OTHER_DELAY' en nuestro data frame cuyo valor sea igual al delay

In [20]:
# Creamos la OTHER_DELAY con los mismos datos que ARRIVAL_DELAY
df['OTHER_DELAY'] = df['ARRIVAL_DELAY']

# Como hemos visto que si una columna _DELAY es NaN el resto también, utilizamos una única columa para comparar, y asignamos a OTHER_DELAY la diferencia entre el delay a la llegada y el resto de delays
df.loc[pd.notna(df['AIR_SYSTEM_DELAY']),'OTHER_DELAY'] = df['ARRIVAL_DELAY'] - df['AIR_SYSTEM_DELAY'] - df['SECURITY_DELAY'] - df['AIRLINE_DELAY'] -df['LATE_AIRCRAFT_DELAY'] - df['WEATHER_DELAY']
df[45:50]

DATE AIRLINE FLIGHT_NUMBER ORIGIN_AIRPORT DESTINATION_AIRPORT  \
194 2015-01-01      UA          1198            BOS                 ORD   
195 2015-01-01      UA          1224            SFO                 LAX   
197 2015-01-01      UA          1431            BOS                 LAX   
199 2015-01-01      UA          1735            LAX                 ORD   
201 2015-01-01      US           459            MCO                 PHX   

     SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY  TAXI_OUT  \
194                  600           557.0             -3.0      13.0   
195                  600           605.0              5.0      11.0   
197                  600           603.0              3.0      13.0   
199                  600           607.0              7.0      13.0   
201                  600           557.0             -3.0      13.0   

     WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  WHEELS_ON  \
194       610.0           173.0         165.0     146.0       867      736.0   
195       616.0            88.0          72.0      52.0       337      708.0   
197       616.0           404.0         396.0     376.0      2611      932.0   
199       620.0           250.0         227.0     201.0      1744     1141.0   
201       610.0           281.0         294.0     273.0      1849      843.0   

     TAXI_IN  SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  \
194      6.0                753         742.0          -11.0   
195      9.0                728         717.0          -11.0   
197      7.0                944         939.0           -5.0   
199     13.0               1210        1154.0          -16.0   
201      8.0                841         851.0           10.0   

     AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
194               NaN             NaN            NaN                  NaN   
195               NaN             NaN            NaN                  NaN   
197               NaN             NaN            NaN                  NaN   
199               NaN             NaN            NaN                  NaN   
201               NaN             NaN            NaN                  NaN   

     WEATHER_DELAY  OTHER_DELAY  
194            NaN        -11.0  
195            NaN        -11.0  
197            NaN         -5.0  
199            NaN        -16.0  
201            NaN         10.0

In [21]:
# Cambiamos OTHEY_DELAY <0 por 0 dado que estos vuelos han llegado antes de lo previsto, no han experimentado un retraso
df.loc[df["OTHER_DELAY"] < 0, "OTHER_DELAY"] = 0
df[45:50]

DATE AIRLINE FLIGHT_NUMBER ORIGIN_AIRPORT DESTINATION_AIRPORT  \
194 2015-01-01      UA          1198            BOS                 ORD   
195 2015-01-01      UA          1224            SFO                 LAX   
197 2015-01-01      UA          1431            BOS                 LAX   
199 2015-01-01      UA          1735            LAX                 ORD   
201 2015-01-01      US           459            MCO                 PHX   

     SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY  TAXI_OUT  \
194                  600           557.0             -3.0      13.0   
195                  600           605.0              5.0      11.0   
197                  600           603.0              3.0      13.0   
199                  600           607.0              7.0      13.0   
201                  600           557.0             -3.0      13.0   

     WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  WHEELS_ON  \
194       610.0           173.0         165.0     146.0       867      736.0   
195       616.0            88.0          72.0      52.0       337      708.0   
197       616.0           404.0         396.0     376.0      2611      932.0   
199       620.0           250.0         227.0     201.0      1744     1141.0   
201       610.0           281.0         294.0     273.0      1849      843.0   

     TAXI_IN  SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  \
194      6.0                753         742.0          -11.0   
195      9.0                728         717.0          -11.0   
197      7.0                944         939.0           -5.0   
199     13.0               1210        1154.0          -16.0   
201      8.0                841         851.0           10.0   

     AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
194               NaN             NaN            NaN                  NaN   
195               NaN             NaN            NaN                  NaN   
197               NaN             NaN            NaN                  NaN   
199               NaN             NaN            NaN                  NaN   
201               NaN             NaN            NaN                  NaN   

     WEATHER_DELAY  OTHER_DELAY  
194            NaN          0.0  
195            NaN          0.0  
197            NaN          0.0  
199            NaN          0.0  
201            NaN         10.0

In [22]:
# Early arrival flights
early_arrival = df[df['ARRIVAL_DELAY']<=0]

print(len(early_arrival))
early_arrival.isna().sum()

910473


DATE                        0
AIRLINE                     0
FLIGHT_NUMBER               0
ORIGIN_AIRPORT              0
DESTINATION_AIRPORT         0
SCHEDULED_DEPARTURE         0
DEPARTURE_TIME              0
DEPARTURE_DELAY             0
TAXI_OUT                    0
WHEELS_OFF                  0
SCHEDULED_TIME              0
ELAPSED_TIME                0
AIR_TIME                    0
DISTANCE                    0
WHEELS_ON                   0
TAXI_IN                     0
SCHEDULED_ARRIVAL           0
ARRIVAL_TIME                0
ARRIVAL_DELAY               0
AIR_SYSTEM_DELAY       910473
SECURITY_DELAY         910473
AIRLINE_DELAY          910473
LATE_AIRCRAFT_DELAY    910473
WEATHER_DELAY          910473
OTHER_DELAY                 0
dtype: int64

El resto de valores NaN en dichas columnas corresponden a aquellos vuelos que han llegado antes de tiempo a destino, cosa que tiene sentido dado que no han experimentado ningún delay

Una vez analizado el por qué de los valores NaN presentes en el dataset parece razonable sustituir dichos valores por 0

In [23]:
# Asignamos a todos los NaNs el valor 0, ya que ahora todo el retraso de sus vuelos está plasmado en la variable OTHER_DELAY
df = df.fillna(0)
df[45:50]

DATE AIRLINE  FLIGHT_NUMBER ORIGIN_AIRPORT DESTINATION_AIRPORT  \
194 2015-01-01      UA           1198            BOS                 ORD   
195 2015-01-01      UA           1224            SFO                 LAX   
197 2015-01-01      UA           1431            BOS                 LAX   
199 2015-01-01      UA           1735            LAX                 ORD   
201 2015-01-01      US            459            MCO                 PHX   

     SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY  TAXI_OUT  \
194                  600           557.0             -3.0      13.0   
195                  600           605.0              5.0      11.0   
197                  600           603.0              3.0      13.0   
199                  600           607.0              7.0      13.0   
201                  600           557.0             -3.0      13.0   

     WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  WHEELS_ON  \
194       610.0           173.0         165.0     146.0       867      736.0   
195       616.0            88.0          72.0      52.0       337      708.0   
197       616.0           404.0         396.0     376.0      2611      932.0   
199       620.0           250.0         227.0     201.0      1744     1141.0   
201       610.0           281.0         294.0     273.0      1849      843.0   

     TAXI_IN  SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  \
194      6.0                753         742.0          -11.0   
195      9.0                728         717.0          -11.0   
197      7.0                944         939.0           -5.0   
199     13.0               1210        1154.0          -16.0   
201      8.0                841         851.0           10.0   

     AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
194               0.0             0.0            0.0                  0.0   
195               0.0             0.0            0.0                  0.0   
197               0.0             0.0            0.0                  0.0   
199               0.0             0.0            0.0                  0.0   
201               0.0             0.0            0.0                  0.0   

     WEATHER_DELAY  OTHER_DELAY  
194            0.0          0.0  
195            0.0          0.0  
197            0.0          0.0  
199            0.0          0.0  
201            0.0         10.0

fillna() sustituye los NaN de todo el dataframe por lo que es importante destacar que podemos utilizar este método dado que los únicos campos NaN del dataframe se encuentran en las variables de delay analizadas. 

In [24]:
df.isna().sum()

DATE                   0
AIRLINE                0
FLIGHT_NUMBER          0
ORIGIN_AIRPORT         0
DESTINATION_AIRPORT    0
SCHEDULED_DEPARTURE    0
DEPARTURE_TIME         0
DEPARTURE_DELAY        0
TAXI_OUT               0
WHEELS_OFF             0
SCHEDULED_TIME         0
ELAPSED_TIME           0
AIR_TIME               0
DISTANCE               0
WHEELS_ON              0
TAXI_IN                0
SCHEDULED_ARRIVAL      0
ARRIVAL_TIME           0
ARRIVAL_DELAY          0
AIR_SYSTEM_DELAY       0
SECURITY_DELAY         0
AIRLINE_DELAY          0
LATE_AIRCRAFT_DELAY    0
WEATHER_DELAY          0
OTHER_DELAY            0
dtype: int64

Resulta interesante saber si los vuelos pueden retrasarse por un único motivo o exclusivamente por uno. Vamos a investigarlo

In [25]:
# Tomamos la variable WEATHER_DELAY como referencia
weather = df.loc[(df['WEATHER_DELAY'] != df['ARRIVAL_DELAY']) & (df['WEATHER_DELAY']>0 )]
weather.head()

DATE AIRLINE  FLIGHT_NUMBER ORIGIN_AIRPORT DESTINATION_AIRPORT  \
3740  2015-01-01      OO           5295            ORD                 CMX   
9142  2015-01-01      UA           1250            MCO                 DEN   
12607 2015-01-01      AA           1150            ORD                 PHX   
12863 2015-01-01      AA           2353            ORD                 DFW   
17286 2015-01-02      MQ           3388            ORD                 JAX   

       SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY  TAXI_OUT  \
3740                  1025          1145.0             80.0      11.0   
9142                  1610          1618.0              8.0      11.0   
12607                 2005          2024.0             19.0      13.0   
12863                 2030          2150.0             80.0      15.0   
17286                  830           858.0             28.0      21.0   

       WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  \
3740       1156.0            83.0          80.0      66.0       359   
9142       1629.0           245.0         255.0     231.0      1546   
12607      2037.0           230.0         233.0     212.0      1440   
12863      2205.0           145.0         145.0     125.0       802   
17286       919.0           133.0         138.0     113.0       864   

       WHEELS_ON  TAXI_IN  SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  \
3740      1402.0      3.0               1248        1405.0           77.0   
9142      1820.0     13.0               1815        1833.0           18.0   
12607     2309.0      8.0               2255        2317.0           22.0   
12863       10.0      5.0               2255          15.0           80.0   
17286     1212.0      4.0               1143        1216.0           33.0   

       AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
3740                0.0             0.0            0.0                 16.0   
9142               10.0             0.0            0.0                  0.0   
12607               3.0             0.0            0.0                  0.0   
12863               0.0             0.0           61.0                  9.0   
17286               5.0             0.0            0.0                 26.0   

       WEATHER_DELAY  OTHER_DELAY  
3740            61.0          0.0  
9142             8.0          0.0  
12607           19.0          0.0  
12863           10.0          0.0  
17286            2.0          0.0

Confirmamos, el retraso puedes estar asociado a varios motivos

## CHECKPOINT. Datos de vuelos limpios

In [26]:
df.head()

DATE AIRLINE  FLIGHT_NUMBER ORIGIN_AIRPORT DESTINATION_AIRPORT  \
1 2015-01-01      AA           2336            LAX                 PBI   
2 2015-01-01      US            840            SFO                 CLT   
3 2015-01-01      AA            258            LAX                 MIA   
5 2015-01-01      DL            806            SFO                 MSP   
7 2015-01-01      US           2013            LAX                 CLT   

   SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  \
1                   10             2.0             -8.0      12.0        14.0   
2                   20            18.0             -2.0      16.0        34.0   
3                   20            15.0             -5.0      15.0        30.0   
5                   25            20.0             -5.0      18.0        38.0   
7                   30            44.0             14.0      13.0        57.0   

   SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  \
1           280.0         279.0     263.0      2330      737.0      4.0   
2           286.0         293.0     266.0      2296      800.0     11.0   
3           285.0         281.0     258.0      2342      748.0      8.0   
5           217.0         230.0     206.0      1589      604.0      6.0   
7           273.0         249.0     228.0      2125      745.0      8.0   

   SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  AIR_SYSTEM_DELAY  \
1                750         741.0           -9.0               0.0   
2                806         811.0            5.0               0.0   
3                805         756.0           -9.0               0.0   
5                602         610.0            8.0               0.0   
7                803         753.0          -10.0               0.0   

   SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  \
1             0.0            0.0                  0.0            0.0   
2             0.0            0.0                  0.0            0.0   
3             0.0            0.0                  0.0            0.0   
5             0.0            0.0                  0.0            0.0   
7             0.0            0.0                  0.0            0.0   

   OTHER_DELAY  
1          0.0  
2          5.0  
3          0.0  
5          8.0  
7          0.0

## Info Aerolíneas

In [27]:
# Vamos a añadir la el nombre asociado a las airlines
# Cargamos los datos
airlines = pd.read_csv("airlines.csv")
airlines.head()

IATA_CODE                 AIRLINE
0        UA   United Air Lines Inc.
1        AA  American Airlines Inc.
2        US         US Airways Inc.
3        F9  Frontier Airlines Inc.
4        B6         JetBlue Airways

In [28]:
# Vamos a examinar ahora los valores nulos
airlines.isna().sum()

IATA_CODE    0
AIRLINE      0
dtype: int64

In [29]:
# Renombramos la columna "AIRLINE" para poder hacer el join con la tabla de aerolineas
df = df.rename(columns={"AIRLINE": "IATA_CODE"})
df.head()

DATE IATA_CODE  FLIGHT_NUMBER ORIGIN_AIRPORT DESTINATION_AIRPORT  \
1 2015-01-01        AA           2336            LAX                 PBI   
2 2015-01-01        US            840            SFO                 CLT   
3 2015-01-01        AA            258            LAX                 MIA   
5 2015-01-01        DL            806            SFO                 MSP   
7 2015-01-01        US           2013            LAX                 CLT   

   SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  \
1                   10             2.0             -8.0      12.0        14.0   
2                   20            18.0             -2.0      16.0        34.0   
3                   20            15.0             -5.0      15.0        30.0   
5                   25            20.0             -5.0      18.0        38.0   
7                   30            44.0             14.0      13.0        57.0   

   SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  \
1           280.0         279.0     263.0      2330      737.0      4.0   
2           286.0         293.0     266.0      2296      800.0     11.0   
3           285.0         281.0     258.0      2342      748.0      8.0   
5           217.0         230.0     206.0      1589      604.0      6.0   
7           273.0         249.0     228.0      2125      745.0      8.0   

   SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  AIR_SYSTEM_DELAY  \
1                750         741.0           -9.0               0.0   
2                806         811.0            5.0               0.0   
3                805         756.0           -9.0               0.0   
5                602         610.0            8.0               0.0   
7                803         753.0          -10.0               0.0   

   SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  \
1             0.0            0.0                  0.0            0.0   
2             0.0            0.0                  0.0            0.0   
3             0.0            0.0                  0.0            0.0   
5             0.0            0.0                  0.0            0.0   
7             0.0            0.0                  0.0            0.0   

   OTHER_DELAY  
1          0.0  
2          5.0  
3          0.0  
5          8.0  
7          0.0

In [30]:
# Unimos ambas tablas
flights = df.merge(airlines, on='IATA_CODE', how='left')
flights.head()

DATE IATA_CODE  FLIGHT_NUMBER ORIGIN_AIRPORT DESTINATION_AIRPORT  \
0 2015-01-01        AA           2336            LAX                 PBI   
1 2015-01-01        US            840            SFO                 CLT   
2 2015-01-01        AA            258            LAX                 MIA   
3 2015-01-01        DL            806            SFO                 MSP   
4 2015-01-01        US           2013            LAX                 CLT   

   SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  \
0                   10             2.0             -8.0      12.0        14.0   
1                   20            18.0             -2.0      16.0        34.0   
2                   20            15.0             -5.0      15.0        30.0   
3                   25            20.0             -5.0      18.0        38.0   
4                   30            44.0             14.0      13.0        57.0   

   SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  \
0           280.0         279.0     263.0      2330      737.0      4.0   
1           286.0         293.0     266.0      2296      800.0     11.0   
2           285.0         281.0     258.0      2342      748.0      8.0   
3           217.0         230.0     206.0      1589      604.0      6.0   
4           273.0         249.0     228.0      2125      745.0      8.0   

   SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  AIR_SYSTEM_DELAY  \
0                750         741.0           -9.0               0.0   
1                806         811.0            5.0               0.0   
2                805         756.0           -9.0               0.0   
3                602         610.0            8.0               0.0   
4                803         753.0          -10.0               0.0   

   SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  \
0             0.0            0.0                  0.0            0.0   
1             0.0            0.0                  0.0            0.0   
2             0.0            0.0                  0.0            0.0   
3             0.0            0.0                  0.0            0.0   
4             0.0            0.0                  0.0            0.0   

   OTHER_DELAY                 AIRLINE  
0          0.0  American Airlines Inc.  
1          5.0         US Airways Inc.  
2          0.0  American Airlines Inc.  
3          8.0    Delta Air Lines Inc.  
4          0.0         US Airways Inc.

In [31]:
# Ponemos la nueva columna a continuación del código de la aerolinea
cols = flights.columns.tolist()
cols = cols[0:2]+cols[-1:] + cols[2:-1]
flights = flights[cols] 
flights.head()

DATE IATA_CODE                 AIRLINE  FLIGHT_NUMBER ORIGIN_AIRPORT  \
0 2015-01-01        AA  American Airlines Inc.           2336            LAX   
1 2015-01-01        US         US Airways Inc.            840            SFO   
2 2015-01-01        AA  American Airlines Inc.            258            LAX   
3 2015-01-01        DL    Delta Air Lines Inc.            806            SFO   
4 2015-01-01        US         US Airways Inc.           2013            LAX   

  DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY  \
0                 PBI                   10             2.0             -8.0   
1                 CLT                   20            18.0             -2.0   
2                 MIA                   20            15.0             -5.0   
3                 MSP                   25            20.0             -5.0   
4                 CLT                   30            44.0             14.0   

   TAXI_OUT  WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  \
0      12.0        14.0           280.0         279.0     263.0      2330   
1      16.0        34.0           286.0         293.0     266.0      2296   
2      15.0        30.0           285.0         281.0     258.0      2342   
3      18.0        38.0           217.0         230.0     206.0      1589   
4      13.0        57.0           273.0         249.0     228.0      2125   

   WHEELS_ON  TAXI_IN  SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  \
0      737.0      4.0                750         741.0           -9.0   
1      800.0     11.0                806         811.0            5.0   
2      748.0      8.0                805         756.0           -9.0   
3      604.0      6.0                602         610.0            8.0   
4      745.0      8.0                803         753.0          -10.0   

   AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
0               0.0             0.0            0.0                  0.0   
1               0.0             0.0            0.0                  0.0   
2               0.0             0.0            0.0                  0.0   
3               0.0             0.0            0.0                  0.0   
4               0.0             0.0            0.0                  0.0   

   WEATHER_DELAY  OTHER_DELAY  
0            0.0          0.0  
1            0.0          5.0  
2            0.0          0.0  
3            0.0          8.0  
4            0.0          0.0

## Info Aeropuertos

In [32]:
# Vamos a añadir la el nombre asociado a las airlines
# Cargamos los datos
airports = pd.read_csv("airports.csv")
airports.head()

IATA_CODE                              AIRPORT         CITY STATE COUNTRY  \
0       ABE  Lehigh Valley International Airport    Allentown    PA     USA   
1       ABI             Abilene Regional Airport      Abilene    TX     USA   
2       ABQ    Albuquerque International Sunport  Albuquerque    NM     USA   
3       ABR            Aberdeen Regional Airport     Aberdeen    SD     USA   
4       ABY   Southwest Georgia Regional Airport       Albany    GA     USA   

   LATITUDE  LONGITUDE  
0  40.65236  -75.44040  
1  32.41132  -99.68190  
2  35.04022 -106.60919  
3  45.44906  -98.42183  
4  31.53552  -84.19447

In [33]:
# Vamos a examinar ahora los valores nulos
airports.isna().sum()

IATA_CODE    0
AIRPORT      0
CITY         0
STATE        0
COUNTRY      0
LATITUDE     3
LONGITUDE    3
dtype: int64

In [34]:
lat_null = airports[airports['LATITUDE'].isna()]
lat_null

IATA_CODE                                            AIRPORT  \
96        ECP    Northwest Florida Beaches International Airport   
234       PBG                  Plattsburgh International Airport   
313       UST  Northeast Florida Regional Airport (St. August...   

              CITY STATE COUNTRY  LATITUDE  LONGITUDE  
96     Panama City    FL     USA       NaN        NaN  
234    Plattsburgh    NY     USA       NaN        NaN  
313  St. Augustine    FL     USA       NaN        NaN

Falta información de 3 aeropuertos, así que la rellenaremos buscando sus datos en internet e introduciéndola manualmente

In [35]:
ECP_COORD = [30.3549, 85.7995]
PBG_COORD = [44.6521, 73.4679]
UST_COORD = [29.9544, 81.3429]
airports.at[96,["LATITUDE","LONGITUDE"]]= ECP_COORD
airports.at[234,["LATITUDE","LONGITUDE"]]= PBG_COORD
airports.at[313,["LATITUDE","LONGITUDE"]]= UST_COORD
airports.isna().sum()

IATA_CODE    0
AIRPORT      0
CITY         0
STATE        0
COUNTRY      0
LATITUDE     0
LONGITUDE    0
dtype: int64

In [36]:
#lat_null[lat_null.IATA_CODE.isin(top_10_airports)]

**NOTA** No haría falta dado que ninguno de estos aeropuertos pertenece al top 10

In [37]:
# Eliminamos la columna airport, ya que trabajaremos únicamente con los códigos de los aeropuertos, no con sus nombres
airports = airports.drop(columns = "AIRPORT")

In [38]:
# Creamos dos bases de datos para hacer el join con la principal tanto para aeropuertos de llegada como de salida
airports= airports.rename(columns={"IATA_CODE": "AIRPORT"})

origin_airports = airports.add_prefix('ORIGIN_')
destination_airports = airports.add_prefix('DESTINATION_')

origin_airports.head()

ORIGIN_AIRPORT  ORIGIN_CITY ORIGIN_STATE ORIGIN_COUNTRY  ORIGIN_LATITUDE  \
0            ABE    Allentown           PA            USA         40.65236   
1            ABI      Abilene           TX            USA         32.41132   
2            ABQ  Albuquerque           NM            USA         35.04022   
3            ABR     Aberdeen           SD            USA         45.44906   
4            ABY       Albany           GA            USA         31.53552   

   ORIGIN_LONGITUDE  
0         -75.44040  
1         -99.68190  
2        -106.60919  
3         -98.42183  
4         -84.19447

In [39]:
flights.head()

DATE IATA_CODE                 AIRLINE  FLIGHT_NUMBER ORIGIN_AIRPORT  \
0 2015-01-01        AA  American Airlines Inc.           2336            LAX   
1 2015-01-01        US         US Airways Inc.            840            SFO   
2 2015-01-01        AA  American Airlines Inc.            258            LAX   
3 2015-01-01        DL    Delta Air Lines Inc.            806            SFO   
4 2015-01-01        US         US Airways Inc.           2013            LAX   

  DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY  \
0                 PBI                   10             2.0             -8.0   
1                 CLT                   20            18.0             -2.0   
2                 MIA                   20            15.0             -5.0   
3                 MSP                   25            20.0             -5.0   
4                 CLT                   30            44.0             14.0   

   TAXI_OUT  WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  \
0      12.0        14.0           280.0         279.0     263.0      2330   
1      16.0        34.0           286.0         293.0     266.0      2296   
2      15.0        30.0           285.0         281.0     258.0      2342   
3      18.0        38.0           217.0         230.0     206.0      1589   
4      13.0        57.0           273.0         249.0     228.0      2125   

   WHEELS_ON  TAXI_IN  SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  \
0      737.0      4.0                750         741.0           -9.0   
1      800.0     11.0                806         811.0            5.0   
2      748.0      8.0                805         756.0           -9.0   
3      604.0      6.0                602         610.0            8.0   
4      745.0      8.0                803         753.0          -10.0   

   AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
0               0.0             0.0            0.0                  0.0   
1               0.0             0.0            0.0                  0.0   
2               0.0             0.0            0.0                  0.0   
3               0.0             0.0            0.0                  0.0   
4               0.0             0.0            0.0                  0.0   

   WEATHER_DELAY  OTHER_DELAY  
0            0.0          0.0  
1            0.0          5.0  
2            0.0          0.0  
3            0.0          8.0  
4            0.0          0.0

In [40]:
# Unimos las tres tablas
flights_origin = flights.merge(origin_airports, on='ORIGIN_AIRPORT', how='left')
flights_complete = flights_origin.merge(destination_airports, on='DESTINATION_AIRPORT', how='left')
flights_complete.head()

DATE IATA_CODE                 AIRLINE  FLIGHT_NUMBER ORIGIN_AIRPORT  \
0 2015-01-01        AA  American Airlines Inc.           2336            LAX   
1 2015-01-01        US         US Airways Inc.            840            SFO   
2 2015-01-01        AA  American Airlines Inc.            258            LAX   
3 2015-01-01        DL    Delta Air Lines Inc.            806            SFO   
4 2015-01-01        US         US Airways Inc.           2013            LAX   

  DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  DEPARTURE_TIME  DEPARTURE_DELAY  \
0                 PBI                   10             2.0             -8.0   
1                 CLT                   20            18.0             -2.0   
2                 MIA                   20            15.0             -5.0   
3                 MSP                   25            20.0             -5.0   
4                 CLT                   30            44.0             14.0   

   TAXI_OUT  WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  \
0      12.0        14.0           280.0         279.0     263.0      2330   
1      16.0        34.0           286.0         293.0     266.0      2296   
2      15.0        30.0           285.0         281.0     258.0      2342   
3      18.0        38.0           217.0         230.0     206.0      1589   
4      13.0        57.0           273.0         249.0     228.0      2125   

   WHEELS_ON  TAXI_IN  SCHEDULED_ARRIVAL  ARRIVAL_TIME  ARRIVAL_DELAY  \
0      737.0      4.0                750         741.0           -9.0   
1      800.0     11.0                806         811.0            5.0   
2      748.0      8.0                805         756.0           -9.0   
3      604.0      6.0                602         610.0            8.0   
4      745.0      8.0                803         753.0          -10.0   

   AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
0               0.0             0.0            0.0                  0.0   
1               0.0             0.0            0.0                  0.0   
2               0.0             0.0            0.0                  0.0   
3               0.0             0.0            0.0                  0.0   
4               0.0             0.0            0.0                  0.0   

   WEATHER_DELAY  OTHER_DELAY    ORIGIN_CITY ORIGIN_STATE ORIGIN_COUNTRY  \
0            0.0          0.0    Los Angeles           CA            USA   
1            0.0          5.0  San Francisco           CA            USA   
2            0.0          0.0    Los Angeles           CA            USA   
3            0.0          8.0  San Francisco           CA            USA   
4            0.0          0.0    Los Angeles           CA            USA   

   ORIGIN_LATITUDE  ORIGIN_LONGITUDE DESTINATION_CITY DESTINATION_STATE  \
0         33.94254        -118.40807  West Palm Beach                FL   
1         37.61900        -122.37484        Charlotte                NC   
2         33.94254        -118.40807            Miami                FL   
3         37.61900        -122.37484      Minneapolis                MN   
4         33.94254        -118.40807        Charlotte                NC   

  DESTINATION_COUNTRY  DESTINATION_LATITUDE  DESTINATION_LONGITUDE  
0                 USA              26.68316              -80.09559  
1                 USA              35.21401              -80.94313  
2                 USA              25.79325              -80.29056  
3                 USA              44.88055              -93.21692  
4                 USA              35.21401              -80.94313

In [41]:
len(flights_complete)

1463859

**NOTA:** La info relativa a los aeropuertos la añadiremos más adelante cuando estemos trabajando con los aeropuertos

In [42]:
# Guardamos los datos preprocesados, para ser utilizados en la predicción posterior
flights_complete.to_csv("flightsCleaned.csv", index=False)